File: components/schemes/State/examples/StateSchemeExampleController.m
// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "StateSchemeExampleController.h"
#import "MaterialPalettes.h"
#import "MaterialStates.h"

@interface StateSchemeExampleController ()

@property(nullable, strong, nonatomic) MDCStateExampleColorScheme *exampleColorScheme;
@property(nullable, strong, nonatomic) MDCStateScheme *stateScheme;

@end

@implementation StateSchemeExampleController

- (void)viewDidLoad {
  [super viewDidLoad];

  // Note: State and Color schemes should be initialized with the same defaults, to ensure
  //       State overrides match the colors in the color scheme.
  if (self.exampleColorScheme == nil) {
    self.exampleColorScheme =
        [[MDCStateExampleColorScheme alloc] initWithDefaults:MDCColorSchemeDefaultsMaterial201804];
  }
  if (self.stateScheme == nil) {
    self.stateScheme =
        [[MDCStateScheme alloc] initWithDefaults:MDCColorSchemeDefaultsMaterial201804];
  }

  self.view.backgroundColor = self.exampleColorScheme.backgroundColor;

  // Note: There is no UI for this example.
  [self runStateSchemeExperiment];
}

/**
 The following retrieves various state colors for the OnWhite theme.

 The console output after running this experiment example is:

 State Example> onWhiteNormalOverlayColor should be: [null]. Is it? [YES]
 State Example> onWhiteSelectedOverlayColor should be: [overlay-12%]. Is it? [YES]
 State Example> overlay for state: [NORMAL] is: [(null)]. It should be: [null]
 State Example> overlay for state: [PRESSED] is:
                [semantic: overlay | palette: --, tint: --, hex: --, opacity: 0.10, dim: 1.00].
                It should be: [overlay]
 State Example> content for state: [DISABLED] is:
                [semantic: -- | palette: grey, tint: 800, hex: --, opacity: 0.38, dim: 1.00]
                It should be: [Palette.grey.tint800 at 38%]. Is it? [YES]
 State Example> content for state: [PRESSED] is:
                [semantic: on-surface | palette: --, tint: --, hex: --, opacity: 1.00, dim: 0.20].
                It should be: [on-surface 20% dim]
 State Example> overriding the content color for the PRESSED state to: Primary Color
 State Example> modified content for state: [PRESSED] is:
                [semantic: primary | palette: --, tint: --, hex: --, opacity: 0.60, dim: 1.00].
                It should be: [PrimaryColor 60%]

*/
- (void)runStateSchemeExperiment {
  UIColor *onWhiteNormalOverlayColor =
      [[self.stateScheme overlayColorResourceForState:MDCControlStateNormal
                                                theme:MDCStateSchemeThemeOnWhite]
          colorWithColorScheme:self.exampleColorScheme];
  NSLog(@"State Example> onWhiteNormalOverlayColor should be: [null]. Is it? [%@]",
        onWhiteNormalOverlayColor == nil ? @"YES" : @"NO");

  UIColor *onWhiteSelectedOverlayColor =
      [[self.stateScheme overlayColorResourceForState:MDCControlStateSelected
                                                theme:MDCStateSchemeThemeOnWhite]
          colorWithColorScheme:self.exampleColorScheme];

  NSLog(@"State Example> onWhiteSelectedOverlayColor should be: [overlay-12%%]. Is it? [%@]",
        [onWhiteSelectedOverlayColor
            isEqual:[self.exampleColorScheme.overlayColor colorWithAlphaComponent:0.12]]
            ? @"YES"
            : @"NO");

  MDCColorResource *normalOverlay =
      [self.stateScheme overlayColorResourceForState:MDCControlStateNormal
                                               theme:MDCStateSchemeThemeOnWhite];
  NSLog(@"State Example> overlay for state: [NORMAL] is: [%@]. It should be: [null]",
        normalOverlay);

  MDCColorResource *pressedOverlay =
      [self.stateScheme overlayColorResourceForState:MDCControlStateHighlighted
                                               theme:MDCStateSchemeThemeOnWhite];
  NSLog(@"State Example> overlay for state: [PRESSED] is: [%@]. It should be: [overlay]",
        pressedOverlay);

  MDCColorResource *disabledContent =
      [self.stateScheme contentColorResourceForState:MDCControlStateDisabled
                                               theme:MDCStateSchemeThemeOnWhite];
  NSLog(@"State Example> content for state: [DISABLED] is: [%@]", disabledContent);
  if (disabledContent != nil) {
    UIColor *disabledColor = [disabledContent colorWithColorScheme:self.exampleColorScheme];
    UIColor *intendedDisabledColor = [MDCPalette.greyPalette.tint800 colorWithAlphaComponent:0.38];
    NSLog(@"State Example> It should be: [Palette.grey.tint800 at 38%%]. Is it? [%@]",
          [disabledColor isEqual:intendedDisabledColor] ? @"YES" : @"NO");
  }

  MDCColorResource *pressedContent =
      [self.stateScheme contentColorResourceForState:MDCControlStateHighlighted
                                               theme:MDCStateSchemeThemeOnWhite];
  NSLog(
      @"State Example> content for state: [PRESSED] is: [%@]. It should be: [on-surface 20%% dim]",
      pressedContent);

  // override the content color for the pressed state
  [self.stateScheme setContentColorWithSemanticColor:MDCColorResourceSemanticPrimary
                                             opacity:0.6
                                            forState:MDCControlStateInteractive
                                             inTheme:MDCStateSchemeThemeOnWhite];
  NSLog(@"State Example> overriding the content color for the PRESSED state to: Primary Color");
  MDCColorResource *newPressedContent =
      [self.stateScheme contentColorResourceForState:MDCControlStateHighlighted
                                               theme:MDCStateSchemeThemeOnWhite];
  NSLog(@"State Example> modified content for state: [PRESSED] is: [%@].", newPressedContent);
  NSLog(@"It should be: [PrimaryColor 60%%]");
}

@end

// MARK: Catalog by Convention

@implementation StateSchemeExampleController (CatalogByConvention)

+ (NSDictionary *)catalogMetadata {
  return @{
    @"breadcrumbs" : @[ @"Schemes", @"States Prototype (Alpha)" ],
    @"description" : @"Prototype of States Theming.",
    @"primaryDemo" : @NO,
    @"presentable" : @NO,
  };
}

@end

File: components/private/State/examples/StateSchemeExampleController.swift
// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import UIKit

class StateSchemeExampleController: UIViewController {

  let colorScheme = StateExampleColorScheme(withDefaults: .material201804)
  let stateScheme = StateScheme(withDefaults: .material201804)

  override func viewDidLoad() {
    super.viewDidLoad()

    self.view.backgroundColor = colorScheme.backgroundColor

    experimentWithSchemeAPI()
  }

  /*
   The following retrieves the overlay and content colors for the normal state in the on-white theme.

   The console output after running this example is:

   State Example> overlayColorResource for state: [enabled] is: [ nil ]. It should be: [nil]

   State Example> contentColorResource for state: [disabled] is: [palette: grey tint: tint800 opacity: 0.38]
   State Example> It should be: [MDCPalette.grey.tint800 at 38%]. Is it ? true

   State Example> ContentColorResource for state: [pressed] is: [palette: grey tint: tint900]
   State Example> It is overriden by the current color scheme and should therefore be: [MDCPalette.grey.tint900]. Is it ? true

   State Example> contentColorResource for state: [enabled] is: [ semantic: onSurface (on-surface) ]
   State Example> It should be the semantic color: [onSurface]. Is it ? true

   Note: There is no UI for this example.

   */
  func experimentWithSchemeAPI() {

    let enabled = MDCControlState.normal
    let disabled = MDCControlState.disabled
    let pressed = MDCControlState.highlighted

    let overlayResource = stateScheme.overlayColorResource(forState: enabled, forTheme: StateTheme.onWhite)
    print("State Example> overlayColorResource for state: [\(enabled)] is: [", overlayResource ?? "nil", "]. It should be: [nil]")

    if let disabledResource = stateScheme.contentColorResource(forState: disabled, forTheme: StateTheme.onWhite) {
      print("State Example> contentColorResource for state: [\(disabled)] is: [\(disabledResource)]")
      let disabledColor = colorScheme.realColor(colorResource: disabledResource)
      print("State Example> It should be: [MDCPalette.grey.tint800 at 38%]. Is it ?", disabledColor == MDCPalette.grey.tint800.withAlphaComponent(0.38))
    } else {
      print("State Example> Error: disabledResource should not be nil")
    }

    if let pressedResource = stateScheme.contentColorResource(forState: pressed, forTheme: StateTheme.onWhite) {
      print("State Example> ContentColorResource for state: [\(pressed)] is: [\(pressedResource)]")
      let pressedColor = colorScheme.realColor(colorResource: pressedResource)
      print("State Example> It is overriden by the current color scheme and should therefore be: [MDCPalette.grey.tint900]. Is it ?", pressedColor == MDCPalette.grey.tint900)
    } else {
      print("State Example> Error: pressedColor should not be nil")
    }

    if let contentResource = stateScheme.contentColorResource(forState: enabled, forTheme: StateTheme.onWhite) {
      print("State Example> contentColorResource for state: [\(enabled)] is: [", contentResource, "]")

      let contentColor = colorScheme.realColor(colorResource: contentResource)
      print("State Example> It should be the semantic color: [onSurface]. Is it ?", contentColor == colorScheme.onSurfaceColor)
    } else {
      print("State Example> Error: contentColor should not be nil")
    }
  }
}

// MARK: Catalog by convention

extension StateSchemeExampleController {
  class func catalogMetadata() -> [String: Any] {
    return [
      "breadcrumbs": ["Alpha: States", "States Prototype"],
      "primaryDemo": false,
      "presentable": false,
    ]
  }
}
